File: doc/tutoNLP/loops/l5_gpu_reduction_nested_write.cpp
// Difficulty: Level 5
// Request: target the write to d_partial_sums inside the bi and ti loops of reduce

void reduce(float* d_partial_sums) {
  for (int bi = 0; bi < 128; bi++) {
    for (int ti = 0; ti < 256; ti++) {
      d_partial_sums[bi] += ti;
    }
  }
}

File: doc/tutoNLP/loops/l5_gpu_reduction_nested_write.ml
open Optitrust
open Prelude

(* Difficulty: Level 5
   Request: target the write to d_partial_sums inside the bi and ti loops of reduce
   Expected target: [cFunDef "reduce"; cFor "bi"; cFor "ti"; cArrayWrite "d_partial_sums"]
   Accepted variant: [cTopFunDef "reduce"; cFor "bi"; cFor "ti"; cArrayWrite "d_partial_sums"] *)

let _ = Run.script_cpp (fun () ->
  !! Show.target [cFunDef "reduce"; cFor "bi"; cFor "ti"; cArrayWrite "d_partial_sums"];
)
